### PR TITLE
HTTP Status, Task publisher

### PR DESCRIPTION
**Pull request by joelrebel**
- Adds an implementation of a HTTP based Status Value publisher.
- Adds an implementation of a HTTP based Task publisher.
- Adds a Publisher interface from a Controller perspective accepts an updated Task object and publishes required Status, Task updates.

Not to be merged until https://github.com/metal-toolbox/conditionorc/pull/228 is merged, released and this branch is pinned on that release.

### CODE REVIEW

File: events/controller/controller_test.go [hunk 1]
```diff
@@ -282,7 +282,12 @@ func TestConditionFromEvent(t *testing.T) {
 
 func TestRunConditionHandlerWithMonitor(t *testing.T) {
 	// test for no leaked go routines
-	defer goleak.VerifyNone(t)
+	defer goleak.VerifyNone(t, []goleak.Option{
+		// Ignore can be removed if this ever gets fixed,
+		//
+		// https://github.com/census-instrumentation/opencensus-go/issues/1191
+		goleak.IgnoreTopFunction("go.opencensus.io/stats/view.(*worker).start"),
```
Comment (DoctorVin) [inline events/controller/controller_test.go]: You might also try `goleak.IgnoreCurrent`.

File: events/controller/publisher.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+func (p *PublisherHTTP) Publish(ctx context.Context, task *condition.Task[any, any], tsUpdate bool) error {
+	err := p.statusValuePublisher.Publish(ctx, task.Server.ID, task.State, task.Status.MustMarshal(), tsUpdate)
+	if err != nil {
+		p.logger.WithError(err).Error("Status Value publish error")
+		return err
+	}
+
+	if tsUpdate {
+		task.UpdatedAt = time.Now()
```
Comment (DoctorVin) [inline events/controller/publisher.go]: This should be done unconditionally.

File: events/controller/status.go [hunk 3]
```diff
@@ -450,3 +453,71 @@ func (p *natsEventStatusAcknowleger) nak() {
...
+	resp, err := s.orcQueryor.ConditionStatusUpdate(ctx, s.conditionKind, s.serverID, s.conditionID, s.controllerID, sv, tsUpdateOnly)
+	if err != nil {
+		s.logger.WithError(err).Error("condition status update error")
+		return err
+	}
+
+	if resp.StatusCode != 200 {
+		// nolint:goerr113 // I'd like to keep the error definition close to where its in use.
+		errNon200 := fmt.Errorf("non 200 response code returned: %d", resp.StatusCode)
+		s.logger.Error(errNon200)
```
Comment (DoctorVin) [inline events/controller/status.go]: What message gets logged here? Is there enough context to debug?

File: events/controller/status.go [hunk 4]
```diff
@@ -450,3 +453,71 @@ func (p *natsEventStatusAcknowleger) nak() {
...
+		WorkerID:  s.controllerID.String(),
+		Target:    serverID,
+		TraceID:   trace.SpanFromContext(ctx).SpanContext().TraceID().String(),
+		SpanID:    trace.SpanFromContext(ctx).SpanContext().SpanID().String(),
+		State:     string(state),
+		Status:    status,
+		UpdatedAt: time.Now(),
+	}
+
+	resp, err := s.orcQueryor.ConditionStatusUpdate(ctx, s.conditionKind, s.serverID, s.conditionID, s.controllerID, sv, tsUpdateOnly)
```
Comment (DoctorVin) [inline events/controller/status.go:503]: `tsUpdateOnly` shouldn't leak into the ConditionOrc API, IMO. The client should update the status value as required and pass that to ConditionOrc to apply as required.

File: events/controller/status.go [hunk 5]
```diff
@@ -450,3 +453,71 @@ func (p *natsEventStatusAcknowleger) nak() {
...
+	if err != nil {
+		s.logger.WithError(err).Error("condition status update error")
+		return err
+	}
+
+	if resp.StatusCode != 200 {
+		// nolint:goerr113 // I'd like to keep the error definition close to where its in use.
+		errNon200 := fmt.Errorf("non 200 response code returned: %d", resp.StatusCode)
```
Comment (DoctorVin) [inline events/controller/status.go]: This is pretty restrictive. A `201` would trip an error?

File: events/controller/task.go [hunk 6]
```diff
@@ -280,3 +283,96 @@ func (n *NatsConditionTaskRepository) Query(ctx context.Context) (*condition.Tas
...
+	serverID,
+	conditionID uuid.UUID,
+	conditionKind condition.Kind,
+	orcQueryor orc.Queryor,
+	logger *logrus.Logger,
+) ConditionTaskRepository {
+	return &HTTPTaskRepository{
+		orcQueryor:    orcQueryor,
+		conditionID:   conditionID,
+		conditionKind: conditionKind,
```
Comment (DoctorVin) [inline events/controller/task.go:306]: why embed the kind on the publisher? A publisher could handle many kinds.

File: events/controller/status.go [hunk 7]
```diff
@@ -450,3 +453,71 @@ func (p *natsEventStatusAcknowleger) nak() {
...
+	conditionID uuid.UUID,
+	conditionKind condition.Kind,
+	controllerID registry.ControllerID,
+	orcQueryor orc.Queryor,
+	logger *logrus.Logger,
+) ConditionStatusPublisher {
+	return &HTTPConditionStatusPublisher{
+		orcQueryor:    orcQueryor,
+		conditionID:   conditionID,
+		conditionKind: conditionKind,
```
Comment (DoctorVin) [inline events/controller/status.go:478]: Why embed the id or the kind here?

File: events/controller/task.go [hunk 8]
```diff
@@ -280,3 +283,96 @@ func (n *NatsConditionTaskRepository) Query(ctx context.Context) (*condition.Tas
...
+	defer span.End()
+
+	resp, err := h.orcQueryor.ConditionTaskPublish(ctx, h.conditionKind, h.serverID, h.conditionID, task, tsUpdateOnly)
+	if err != nil {
+		h.logger.WithError(err).Error("Task update error")
+		return errors.Wrap(errTaskPublish, err.Error())
+	}
+
+	if resp.StatusCode != 200 {
+		// nolint:goerr113 // I'd like to keep the error definition close to where its in use.
```
Comment (jakeschuurmans) [inline events/controller/task.go]: Why though? I dont see any issue with putting this at the top. Also you are creating this multiple times.

**Review by jakeschuurmans — CHANGES_REQUESTED** [review]
Just a small question

File: events/controller/task.go (hunk 8 above)
Comment (jakeschuurmans) [reply events/controller/task.go]: I guess I see the issue with moving this up. I love Go's errors, huge improvement over other langauges, but they are not perfect. Go errors are still kinda messy =/

File: events/controller/publisher.go (hunk 2 above)
Comment (joelrebel) [reply events/controller/publisher.go]: true, this was also redundant

File: events/controller/status.go (hunk 4 above)
Comment (joelrebel) [reply events/controller/status.go:503]: In principle we're in agreement,

Although, the current events/controller library publishes `StatusValue` periodic timestamp updates in a loop as long as the controller handler is executing. Along with that, `StatusValue` updates are published by the controller handler - https://github.com/metal-toolbox/rivets/blob/40f864f79dbbde9e43ddec6163514c429120e487/events/controller/controller.go#L510

I'd like us to come up with a better solution, probably using a channel and a method to write those updates to a shared object, such that the data isn't being clobbered, that way the API only receives an updated object and not the bool.

This is more of a design discussion with changes that are far reaching, we can go over it and have it implemented separately.

File: events/controller/status.go (hunk 3 above)
Comment (joelrebel) [reply events/controller/status.go]: added more context in b64fbdf214736e3ea9b1c256a362317c56d8c0e6

File: events/controller/status.go (hunk 5 above)
Comment (joelrebel) [reply events/controller/status.go]: As of now the API only returns a 200, we cover additional cases as they show up

File: events/controller/task.go (hunk 8 above)
Comment (joelrebel) [reply events/controller/task.go]: fair point, added an error type so its not duplicated [b64fbdf](https://github.com/metal-toolbox/rivets/pull/73/commits/b64fbdf214736e3ea9b1c256a362317c56d8c0e6)

File: events/controller/task.go (hunk 6 above)
Comment (joelrebel) [reply events/controller/task.go:306]: The controller initializes a `Publisher` for each Condition it receives, the `Publisher` is kept around only in the context of the Condition as long as its being processed. This is intentional, especially for how the NATS Publisher is bound to a specific KV key, I don't see any reason to reuse Publishers among Conditions.

**Comment by joelrebel** [comment]
rebased on main

**Review by jakeschuurmans — APPROVED** [review]

File: events/controller/status.go (hunk 4 above)
Comment (DoctorVin) [reply events/controller/status.go:503]: I’m not sure I understand the nuances you see. If you think this is more complicated, I’m fine deferring for now and having a discussion.

**Review by DoctorVin — APPROVED** [review]

**Pull request merged**